libopae/plugins/xfpga/metrics/afu_metrics.c

```diff
@@ -0,0 +1,275 @@
+*/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif // HAVE_CONFIG_H
+
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <dirent.h>
+#include <uuid/uuid.h>
+
+#include "xfpga.h"
+#include "common_int.h"
+#include "metrics_int.h"
+#include "types_int.h"
+#include "opae/metrics.h"
+#include "metrics/vector.h"
+#include "safe_string/safe_string.h"
+
+// AFU BBB GUID
+#define METRICS_BBB_GUID            "87816958-C148-4CD0-9D73-E8F258E9E3D7"
+#define METRICS_BBB_ID_H            0x87816958C1484CD0
+#define METRICS_BBB_ID_L            0x9D73E8F258E9E3D7
+
+#define FEATURE_TYPE_BBB            0x2
+
+#define METRIC_CSR_OFFSET           0x20
+#define METRIC_NEXT_CSR             0x8
+
+
+fpga_result discover_afu_metrics_feature(fpga_handle handle, uint64_t *offset)
+{
+	fpga_result result               = FPGA_OK;
+	feature_definition feature_def;
+	uint64_t bbs_offset              = 0;
+
+	memset_s(&feature_def, 0, sizeof(feature_def));
+
+	if (offset == NULL) {
+		FPGA_ERR("Invlaid Input Paramters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	// Read AFU DFH
+	result = fpgaReadMMIO64(handle, 0, 0x0, &(feature_def.dfh.csr));
+	if (result != FPGA_OK) {
+		FPGA_ERR("Invalid handle file descriptor");
+		result = FPGA_NOT_SUPPORTED;
+		return result;
+	}
+
+	// Serach for AFU Metrics BBB DFH
+	while (feature_def.dfh.eol != 0 && feature_def.dfh.next_header_offset != 0) {
+
+		bbs_offset = feature_def.dfh.next_header_offset;
+
+		result = fpgaReadMMIO64(handle, 0, feature_def.dfh.next_header_offset, &(feature_def.dfh.csr));
+		if (result != FPGA_OK) {
+			FPGA_ERR("Invalid handle file descriptor");
+			result = FPGA_NOT_SUPPORTED;
+			return result;
+		}
+
+		if (feature_def.dfh.type == FEATURE_TYPE_BBB) {
+
+			result = fpgaReadMMIO64(handle, 0, bbs_offset +0x8, &(feature_def.guid[0]));
+			if (result != FPGA_OK) {
+				FPGA_ERR("Invalid handle file descriptor");
+				result = FPGA_NOT_SUPPORTED;
+				return result;
+			}
+
+			result = fpgaReadMMIO64(handle, 0, bbs_offset + 0x10, &(feature_def.guid[1]));
+			if (result != FPGA_OK) {
+				FPGA_ERR("Invalid handle file descriptor");
+				result = FPGA_NOT_SUPPORTED;
+				return result;
+			}
+
+			if (feature_def.guid[0] == METRICS_BBB_ID_L &&
+				feature_def.guid[1] == METRICS_BBB_ID_H) {
+				*offset = bbs_offset;
+				return FPGA_OK;
+			} else	{
+				FPGA_ERR(" Metrics BBB Not Found \n ");
+			}
+
+		}
+
+	}
+
+	FPGA_ERR("AFU Metrics BBB Not Found \n ");
+	return FPGA_NOT_FOUND;
+}
+
+
+fpga_result get_afu_metric_value(fpga_handle handle,
+				fpga_metric_vector *enum_vector,
+				uint64_t metric_num,
+				struct fpga_metric *fpga_metric)
+{
+	fpga_result result                           = FPGA_OK;
+	uint64_t index                               = 0;
+	struct metric_bbb_value metric_csr;
+	struct _fpga_enum_metric *_fpga_enum_metric  = NULL;
+	uint64_t num_enun_metrics                    = 0;
+
+	memset_s(&metric_csr, 0, sizeof(metric_csr));
```

Comment:
The signature of memset_s is..
```c
errno_t memset_s(void *dest, rsize_t len, uint8_t value);
```
This call isn't correct.  Please fix.

Reply:
fixed